### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by roszcz**
Thinking about a better approach to dataset creation, I think this is the data flow we're dealing with:

```mermaid
flowchart TD
    A["HF Dataset: list[MidiPiece]"] -->|augmentation| B["A-Dataset: list[MidiPiece]"]
    B --> |"quantization (bins)"| C["AQ-Dataset: list[MidiPiece]"]
    C --> |"chopping (sequence_length)"| D["CAQ-Dataset: list[MidiPiece]"]
    D --> |src tokenization| E["SourceDataset: list[list[str]]"]
    D --> |tgt tokenization| F["TargetDataset: list[list[str]]"]
    E --> |src_vocab| G("source_token_ids: list[list[int]]")
    F --> |tgt_vocab| H("target_token_ids: list[list[int]]")
```

- In terms of caching, I'd say that when we figure out good augmentation, quantization, and chopping strategies, we should store datasets in huggingface.
- Consider using Dataset.save_to_disk for local cache

---

- [x] created `MidiEncoder` classes that can take care of both tokenization and vocab lookup
- [x] created a new dataset class and new dataset preprocessing pipeline - it should be easier now to trace any sample to its source
- [x] untangle inference logic in dashboards

### CODE REVIEW

**Comment by roszcz** [comment]
@WojciechMat this is still in progress, but I think I'm mostly done with the way data flows in this project - please take a look at the `MyTokenizedMidiDataset` and `MidiEncoder` classes and how their managed and let me know what you think!

File: data/dataset.py [hunk 1]
```diff
@@ -1,188 +1,160 @@
...
+        yield "src_encoder", self.src_encoder
+        yield "tgt_encoder", self.tgt_encoder
+
+    def __getitem__(self, idx: int) -> dict:
+        record = self.dataset[idx]
+        source_tokens_ids = self.src_encoder.encode(record)
+        target_tokens_ids = self.tgt_encoder.encode(record)
+
+        out = {
+            "source_token_ids": torch.tensor(source_tokens_ids, dtype=torch.int64),
```
Comment (WojciechMat) [inline data/dataset.py:104]: A bug :)
```suggestion
            "source_tokens_ids": torch.tensor(source_tokens_ids, dtype=torch.int64),
            "tgt_tokens_ids": torch.tensor(tgt_tokens_ids, dtype=torch.int64)
```

File: data/dataset.py [hunk 2]
```diff
@@ -1,188 +1,160 @@
...
+            "target_token_ids": torch.tensor(target_tokens_ids, dtype=torch.int64),
+        }
+        return out
```
Comment (WojciechMat) [inline data/dataset.py:107]: a bug :)
```suggestion
        out = {
            "source_tokens_ids": torch.tensor(source_tokens_ids, dtype=torch.int64),
            "target_tokens_ids": torch.tensor(target_tokens_ids, dtype=torch.int64),
        }
        return out
```

File: train.py [hunk 3]
```diff
@@ -235,11 +243,10 @@ def val_epoch(
...
     total_dist = 0
 
-    dev = torch.device(device)
-
-    for b in tqdm(dataloader):
-        batch = Batch(b[0], b[1], pad=pad_idx)
-        batch.to(dev)
+    for batch in tqdm(dataloader):
+        src = batch["source_tokens_ids"].to(device)
+        tgt = batch["target_tokens_ids"].to(device)
```
Comment (WojciechMat) [inline train.py]: (from here)

**Review by WojciechMat — CHANGES_REQUESTED** [review]

**Comment by WojciechMat** [comment]
Looks like you untangled it a lot and made it more versatile - I will build dstart prediction functionality again on top of that.
I do not have any objections, the code looks a lot cleaner and from what I can tell it makes sense.
Thank you very much for clearing it up!
(I will wait for you to finish this and do experiments with current version of dstart prediction in the meantime)

File: dashboard.py [hunk 4]
```diff
@@ -59,46 +99,45 @@ def model_predictions_review(cfg: DictConfig):
...
     print("Making predictions ...")
-    for record, sample, idx in zip(records, samples, idxs):
-        result = predict_sample(
-            record=sample,
-            dataset=dataset,
+    for record_id in idxs:
+        # Numpy to int :(
+        record = dataset.get_complete_record(int(record_id))
+        record_source = json.loads(record["source"])
+        src_token_ids = record["source_token_ids"]
```
Comment (WojciechMat) [inline dashboard.py]: "source_token_ids" are also used here, so you can also change just the train.py file

File: predict_piece.py [hunk 5]
```diff
@@ -1,66 +1,79 @@
...
+    # Select one full piece
+    record = hf_dataset[record_id]
+    piece = MidiPiece.from_huggingface(record)
+
+    # And run full pre-processing ...
+    qpiece = quantizer.inject_quantization_features(piece)
+    sequences = quantized_piece_to_records(
+        piece=qpiece,
+        sequence_len=train_cfg.dataset.sequence_len,
+        sequence_step=train_cfg.dataset.sequence_len,
```
Comment (WojciechMat) [inline predict_piece.py]: possibly a bug?
```suggestion
        sequence_step=train_cfg.dataset.sequence_step,
```

File: train.py (hunk 3 above)
Comment (WojciechMat) [reply train.py]: PS. you also use correct "source_token_ids" in dashboard.py and predict_piece.py

File: predict_piece.py (hunk 5 above)
Comment (roszcz) [reply predict_piece.py]: I changed that name in the config as well :)

File: dashboard.py (hunk 4 above)
Comment (roszcz) [reply dashboard.py]: I wanted to have `..._token_ids` everywhere, but I made some typos :(

Comment (WojciechMat) [reply dashboard.py]: okay!!

File: predict_piece.py (hunk 5 above)
Comment (WojciechMat) [reply predict_piece.py]: I meant smth else, but now I realize what was the idea, nevermind :))

File: data/dataset.py (hunk 2 above)
Comment (roszcz) [reply data/dataset.py:107]: I'm going with "token ids" since it looks like better english :)

**Review by WojciechMat — APPROVED** [review]

**Pull request merged**